### PR TITLE
Fix #6633: Cargo monitor industry delivery now accounts for which IndustryID the cargo was delivered to

### LINKED ISSUE #6633: cargomonitor industry delivery
**_Samu_ opened the ticket and wrote:**

> In the screenshot BusyBee tells me to deliver wood to the Sawmill to the left.
> I built a Sawmill next to it, to the right, then sent a train delivering wood to both Sawmills.
> 
> The station is closest to the Sawmill to the right, and when inspecting both Sawmill, I see it being to the Sawmill on the right.
> BusyBee, however, progresses my goal, believing I delivered wood to the Sawmill on the left.
> 
> I believe the bug is in cargomonitor.cpp. AddCargoMonitor / industry delivery
> 
> I attached my proposal for a fix.

**Attachments**
- [cargomonitor industry delivery fix v1 r27929.patch](https://bugs.openttd.org/task/6633/getfile/10869/cargomonitor+industry+delivery+fix+v1+r27929.patch) (0.81 KiB)
- [Freningville Transport, 1986-06-21.png](https://bugs.openttd.org/task/6633/getfile/10870/Freningville+Transport%2C+1986-06-21.png) (155.13 KiB)

**Reported version**: _trunk_
**Operating system**: _Windows_

---

### PR DESCRIPTION
**Pull request by SamuXarick**

### CODE REVIEW

**Comment by PeterN** [comment]
This feels wrong, somehow, but I'm not sure how yet.

File: src/cargomonitor.cpp [hunk 1]
```diff
@@ -153,7 +154,7 @@ void AddCargoDelivery(CargoID cargo_type, CompanyID company, uint32 amount, Sour
 	for (const Industry * const *ip = st->industries_near.Begin(); ip != st->industries_near.End(); ip++) {
 		CargoMonitorID num = EncodeCargoIndustryMonitor(company, cargo_type, (*ip)->index);
 		CargoMonitorMap::iterator iter = _cargo_deliveries.find(num);
-		if (iter != _cargo_deliveries.end()) iter->second += amount;
+		if (iter != _cargo_deliveries.end() && (*ip)->index == dest) iter->second += amount;
```
Comment (michicc) [inline src/cargomonitor.cpp]: Make it an `if != dest continue` at the start of the loop? No need to check or find anything just to skip it.

File: src/economy.cpp [hunk 2]
```diff
@@ -1090,30 +1094,30 @@ static Money DeliverGoods(int num_pieces, CargoID cargo_type, StationID dest, Ti
...
-	st->town->received[cs->town_effect].new_act += accepted;
+	st->town->received[cs->town_effect].new_act += accepted_total;
 
 	/* Determine profit */
-	Money profit = GetTransportedGoodsIncome(accepted, DistanceManhattan(source_tile, st->xy), days_in_transit, cargo_type);
+	Money profit = GetTransportedGoodsIncome(accepted_total, DistanceManhattan(source_tile, st->xy), days_in_transit, cargo_type);
 
 	/* Update the cargo monitor. */
-	AddCargoDelivery(cargo_type, company->index, accepted, src_type, src, st);
+	AddCargoDelivery(cargo_type, company->index, num_pieces - accepted_ind, src_type, src, st);
```
Comment (michicc) [inline src/economy.cpp]: If it is a cargo that is not always accepted, the old code would only add the accepted cargo to both the industry and the town cargo monitor.

The new code OTOH will still add the non-accepted amount to the town monitor.

**Review by michicc — DISMISSED** [review]

**Comment by michicc** [comment]
I do think this is a worthwhile change as the description of the cargo monitor API clearly states that deliveries to a certain industry is monitored and not to any industries that happen to be around.

**Comment by TrueBrain** [comment]
@PeterN : ever found out what feels wrong? Otherwise I suggest we go with @michicc , and approve it? :)

**Review by michicc — APPROVED** [review]

**Pull request merged**